File: module_base/src/main/java/com/app/base/web/AppWebViewClient.java
package com.app.base.web;

import android.os.Build;
import android.support.annotation.RequiresApi;
import android.webkit.WebResourceError;
import android.webkit.WebResourceRequest;
import android.webkit.WebView;
import android.webkit.WebViewClient;

/**
 * @author Ztiany
 *         Email: 1169654504@qq.com
 *         Date : 2017-12-20 15:43
 */
abstract class AppWebViewClient extends WebViewClient {


    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    @Override
    public boolean shouldOverrideUrlLoading(WebView view, WebResourceRequest request) {
        String url = request.getUrl().toString();
        return appShouldOverrideUrlLoading(view, url);
    }

    @Override
    public boolean shouldOverrideUrlLoading(WebView view, String url) {
        return appShouldOverrideUrlLoading(view, url);
    }

    abstract boolean appShouldOverrideUrlLoading(WebView view, String url);

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    @Override
    public void onReceivedError(WebView view, WebResourceRequest request, WebResourceError error) {
        super.onReceivedError(view, request, error);
        onAppPageError(request.getUrl().toString());
    }

    @Override
    public void onReceivedError(WebView view, int errorCode, String description, String failingUrl) {
        super.onReceivedError(view, errorCode, description, failingUrl);
       onAppPageError(failingUrl);
    }

    abstract void onAppPageError(String s);

}

File: common_sdk/src/tools-web/java/com/app/base/common/web/AppWebViewClient.kt
package com.app.base.common.web

import android.net.http.SslError
import android.os.Build
import android.webkit.*
import androidx.annotation.RequiresApi
import timber.log.Timber

/**
 * @author Ztiany
 * Date : 2017-12-20 15:43
 */
internal abstract class AppWebViewClient : WebViewClient() {

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    override fun shouldOverrideUrlLoading(view: WebView, request: WebResourceRequest): Boolean {
        Timber.d("shouldOverrideUrlLoading() called with: view = $view, request = $request")
        val url = request.url.toString()
        return appShouldOverrideUrlLoading(view, url)
    }

    override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean {
        Timber.d("shouldOverrideUrlLoading() called with: view = [$view], url = [$url]")
        return appShouldOverrideUrlLoading(view, url)
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    override fun onReceivedError(view: WebView, request: WebResourceRequest, error: WebResourceError) {
        super.onReceivedError(view, request, error)
        onAppPageError(request.url.toString(), error.errorCode, request.isForMainFrame)
    }

    override fun onReceivedError(view: WebView, errorCode: Int, description: String, failingUrl: String) {
        super.onReceivedError(view, errorCode, description, failingUrl)
        Timber.d("onReceivedError() called with: view = [$view], errorCode = [$errorCode], description = [$description], failingUrl = [$failingUrl]")
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            return
        }
        onAppPageError(failingUrl, errorCode, true)
    }

    @RequiresApi(api = Build.VERSION_CODES.M)
    override fun onReceivedHttpError(view: WebView, request: WebResourceRequest, errorResponse: WebResourceResponse) {
        super.onReceivedHttpError(view, request, errorResponse)
        Timber.d("onReceivedHttpError url = %s, isMain = %b, code = %d", request.url.toString(), request.isForMainFrame, errorResponse.statusCode)
        onAppPageError(request.url.toString(), -1, request.isForMainFrame)
    }

    override fun onReceivedSslError(view: WebView, handler: SslErrorHandler, error: SslError) {
        super.onReceivedSslError(view, handler, error)
        Timber.d("onReceivedSslError, url = %s", view.url)
    }

    abstract fun onAppPageError(url: String, code: Int, isMainFrame: Boolean)

    abstract fun appShouldOverrideUrlLoading(view: WebView, url: String): Boolean

}